linsolve/linsolve.py

```diff
@@ -480,27 +544,37 @@ def __init__(self, data, wgts={}, sparse=False, **kwargs):
             try: logampw[eqamp],logphsw[eqphs] = wgts[k], wgts[k]
             except(KeyError): pass
         constants = kwargs.pop('constants', kwargs)
+        self.dtype = infer_dtype(list(data.values()) + list(constants.values()) + list(wgts.values()))
+        self.is_single = self.dtype.name in ['float32', 'complex64']
         logamp_consts, logphs_consts = {}, {}
         for k in constants:
             c = np.log(constants[k]) # log unwraps complex circle at -pi
             logamp_consts[k], logphs_consts[k] = c.real, c.imag
         self.ls_amp = LinearSolver(logamp, logampw, sparse=sparse, constants=logamp_consts)
         self.ls_phs = LinearSolver(logphs, logphsw, sparse=sparse, constants=logphs_consts)
 
-    def solve(self, rcond=1e-10, verbose=False):
+    def solve(self, rcond=1e-10, mode='default', verbose=False):
         """Solve both amplitude and phase by taking the log of both sides to linearize.
 
         Args:
             rcond: cutoff ratio for singular values useed in numpy.linalg.lstsq, numpy.linalg.pinv,
                 or (if sparse) as atol and btol in scipy.sparse.linalg.lsqr
+            mode: 'default', 'lsqr', 'pinv', or 'solve', selects what inverter to use.  'default'
+                is best for most cases, but 'solve' may be faster for fully-constrained systems,
+                and 'pinv' can sometimes be more stable (but slower) than 'lsqr'.
+            verbose: print information about iterations
 
         Returns:
             sol: a dictionary of complex solutions with variables as keys
         """
-        sol_amp = self.ls_amp.solve(rcond=rcond, verbose=verbose)
-        sol_phs = self.ls_phs.solve(rcond=rcond, verbose=verbose)
+        sol_amp = self.ls_amp.solve(rcond=rcond, mode=mode, verbose=verbose)
+        sol_phs = self.ls_phs.solve(rcond=rcond, mode=mode, verbose=verbose)
         sol = {}
-        for k in sol_amp: sol[k] = np.exp(sol_amp[k] + 1j*sol_phs[k])   
+        for k in sol_amp:
+            with warnings.catch_warnings():
```

Comment:
What is this for? At least add a comment explaining what these warnings do and why they can be ignored.

Reply:
Added comment